lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+		return nil, errors.New("storeValidateFailedBlockInBlockIndex: can't set BlockNode to validate failed after it's already validated")
+	}
+	blockNode.Status |= StatusBlockValidateFailed
+	// If the BlockNode is not already stored, we should set it to stored.
+	if !blockNode.IsStored() {
+		blockNode.Status |= StatusBlockStored
+	}
+	// If the DB update fails, then we should return an error.
+	if err = bc.upsertBlockAndBlockNodeToDB(block, blockNode, false); err != nil {
+		return nil, errors.Wrapf(err, "storeValidateFailedBlockInBlockIndex: Problem upserting block and block node to DB")
+	}
+	return blockNode, nil
+}
+
+// storeCommittedBlockInBlockIndex upserts the blocks into the in-memory block index and updates its status to
+// StatusBlockCommitted. If the BlockNode does not have StatusBlockValidated and StatusBlockStored statuses,
+// we also add those. It also writes the block to the block index in badger by calling upsertBlockAndBlockNodeToDB.
+func (bc *Blockchain) storeCommittedBlockInBlockIndex(block *MsgDeSoBlock) (*BlockNode, error) {
+	blockNode, err := bc.getOrCreateBlockNodeFromBlockIndex(block)
+	if err != nil {
+		return nil, errors.Wrapf(err, "storeCommittedBlockInBlockIndex: Problem getting or creating block node")
+	}
+	// If the block is committed, then this is a no-op.
+	if blockNode.IsCommitted() {
+		return blockNode, nil
+	}
+	blockNode.Status |= StatusBlockCommitted
+	if !blockNode.IsValidated() {
+		blockNode.Status |= StatusBlockValidated
+	}
+	if !blockNode.IsStored() {
+		blockNode.Status |= StatusBlockStored
+	}
+	// If the DB update fails, then we should return an error.
+	if err = bc.upsertBlockAndBlockNodeToDB(block, blockNode, true); err != nil {
+		return nil, errors.Wrapf(err, "storeCommittedBlockInBlockIndex: Problem upserting block and block node to DB")
+	}
+	return blockNode, nil
+}
+
+// upsertBlockAndBlockNodeToDB writes the BlockNode to the blockIndexByHash in badger and writes the full block
+// to the db under the <blockHash> -> <serialized block> index.
+func (bc *Blockchain) upsertBlockAndBlockNodeToDB(block *MsgDeSoBlock, blockNode *BlockNode, storeFullBlock bool) error {
+	// Store the block in badger
+	err := bc.db.Update(func(txn *badger.Txn) error {
+		if bc.snapshot != nil {
+			bc.snapshot.PrepareAncestralRecordsFlush()
+			defer bc.snapshot.StartAncestralRecordsFlush(true)
+			glog.V(2).Infof("upsertBlockAndBlockNodeToDB: Preparing snapshot flush")
+		}
+		// TODO: Do we want to write the full block once.
+		// Store the new block in the db under the
+		//   <blockHash> -> <serialized block>
+		// index.
+		// TODO: In the archival mode, we'll be setting ancestral entries for the block reward. Note that it is
+		// 	set in PutBlockWithTxn. Block rewards are part of the state, and they should be identical to the ones
+		// 	we've fetched during Hypersync. Is there an edge-case where for some reason they're not identical? Or
+		// 	somehow ancestral records get corrupted?
```

Comment:
This is the same comment that is in `ProcessBlockPoW` so was leaving it here in case we need to address it here. @AeonSw4n left the original comment in blockchain.go when doing hypersync I believe, so maybe he has more context.

Reply:
I'm not sure if this is what you're referring to, but HyperSync could get messed up if you consider stored blocks as a part of your state. In particular, you can have:
- node A saw block B1 and stored it
- node B did not see block B1 and therefore didn't store it

The solution to the above is to only include data corresponding to the main chain in HyperSync. If you also include data from stored or invalid blocks, then the state that one is hypersyncing can differ. This makes me think we should probably remove the PrepareAncestralRecordsFlush from everywhere EXCEPT for where we COMMIT the block.

Reply:
So PutBlockWithTxn puts block rewards into an index that has `is_state` set to true. So this makes me think that what you're saying is correct. Block rewards from blocks that don't make it into the main chain shouldn't be included. Let's move this discussion offline.

Reply:
resolved in #802